### PR TITLE
Fix compile error when Azure SDK is targeting .NET 6.

### PR DESCRIPTION
**Pull request by AraHaan**
# Description

When I tried to do https://github.com/Azure/azure-sdk-for-net/pull/28911 which basically adds .NET 6 to the .NET Azure SDK, it resulted in compile errors due to the code in this repository. As such this change is a prerequisite of that pr.

# Checklist

To ensure a quick review and merge, please ensure:
- [x] The PR has a understandable title and description explaining the _why_ and _what_.
- [x] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [x] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [x] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

**Comment by AraHaan** [comment]
cc: @annelo-msft for reviews.

File: src/assets/Generator.Shared/StringRequestContent.cs
Comment (AraHaan) [inline src/assets/Generator.Shared/StringRequestContent.cs:23]: I am not sure if this overload of WriteAsync exists in Stream for all of the TFMs used in the .NET Azure SDK repo so I conditioned it like this.

File: src/assets/Generator.Shared/FormUrlEncodedContent.cs
Comment (AraHaan) [inline src/assets/Generator.Shared/FormUrlEncodedContent.cs:38]: I am not sure if this overload of WriteAsync exists in Stream for all of the TFMs used in the .NET Azure SDK repo so I conditioned it like this.

Comment (AraHaan) [reply src/assets/Generator.Shared/FormUrlEncodedContent.cs:38]: It looks like it does exist in TFM’s that implement .NET Standard 2.1.
2FD1335B-29A1-4ADB-B86F-6CE0415E9D8B

File: src/assets/Generator.Shared/StringRequestContent.cs
Comment (AraHaan) [reply src/assets/Generator.Shared/StringRequestContent.cs:23]: It looks like it does exist in TFM’s that implement .NET Standard 2.1.
2FD1335B-29A1-4ADB-B86F-6CE0415E9D8B

File: src/assets/Generator.Shared/StringRequestContent.cs
Comment (AraHaan) [inline src/assets/Generator.Shared/StringRequestContent.cs:22]:
```suggestion
#if !NETFRAMEWORK | !NETSTANDARD2_0
```

File: src/assets/Generator.Shared/FormUrlEncodedContent.cs
Comment (AraHaan) [inline src/assets/Generator.Shared/FormUrlEncodedContent.cs:37]:
```suggestion
#if !NETFRAMEWORK | !NETSTANDARD2_0
```

**Comment by AraHaan** [comment]
I have made changes above to use the span version of WriteAsync where the framework provides it for Stream when it implements .NET Standard 2.1. Since Core > 2.1 implements it and that the Azure SDK targets .NET Framework, Standard 2.0, Core 3.1, and now .NET 6 I think it should be fine after this.

**Review by annelo-msft — COMMENTED** [review]
These are Azure.Core types, one of @christothes, @JoshLove-msft, or myself will need to approve them.  @AlexanderSher FYI.

File: src/assets/Generator.Shared/FormUrlEncodedContent.cs
Comment (AraHaan) [inline src/assets/Generator.Shared/FormUrlEncodedContent.cs:37]:
```suggestion
#if NETSTANDARD2_1 || NETCOREAPP2_1_OR_GREATER
```

File: src/assets/Generator.Shared/StringRequestContent.cs
Comment (AraHaan) [inline src/assets/Generator.Shared/StringRequestContent.cs:22]:
```suggestion
#if NETSTANDARD2_1 || NETCOREAPP2_1_OR_GREATER
```

**Comment by AraHaan** [comment]
I found more that can use System.Memory as well (``Stream.Write()``).

**Comment by AraHaan** [comment]
This should be ready for review, just waiting on when you all have time to review.

**Comment by AraHaan** [comment]
Still working on this.

**Comment by AraHaan** [comment]
cc: @annelo-msft any thoughts on anything else that may cause compile errors when targeting .NET 6 on the Azure SDK?

**Comment by ArthurMa1978** [comment]
Thank you @AraHaan  for your contribution, let me ping @christothes, @AlexanderSher, @annelo-msft, @JoshLove-msft , @m-nash @jsquire .

File: src/AutoRest.CSharp/AutoRest.CSharp.csproj
Comment (jsquire) [reply src/AutoRest.CSharp/AutoRest.CSharp.csproj:28]: +1 to this.

**Review by jsquire — COMMENTED** [review]
Thanks for your contribution, @AraHaan!   

The changes look good to me.  I'll defer to @m-nash for final approval.

File: src/AutoRest.CSharp/AutoRest.CSharp.csproj
Comment (AraHaan) [reply src/AutoRest.CSharp/AutoRest.CSharp.csproj:28]:
```suggestion
    <PackageReference Include="System.Memory" Version="4.5.5" />
```

**Comment by AraHaan** [comment]
I have made the requested changes.

**Comment by archerzz** [comment]
@AraHaan Sorry for the late response. This PR is different after a series of changes. I have a question: why `await stream.WriteAsync(_bytes, 0, _bytes.Length, cancellation).ConfigureAwait(false);` doesn't compile under .Net 6? I checked the [API doc](await stream.WriteAsync(_bytes, 0, _bytes.Length, cancellation).ConfigureAwait(false);) and it should still work under .Net 6.

Is there any benefit to use `await stream.WriteAsync(_bytes.AsMemory(), cancellation).ConfigureAwait(false);`? The former one seems more performant. Thanks.

**Comment by AraHaan** [comment]
The former still compiles, but it generates Visual Studio warnings about using the span versions instead of them (and if warnings as error is enabled it ends up emitting errors instead).

**Review by m-nash — CHANGES_REQUESTED** [review]

File: src/AutoRest.CSharp/AutoRest.CSharp.csproj
Comment (AraHaan) [reply src/AutoRest.CSharp/AutoRest.CSharp.csproj:31]: I think it would be best to consider removing it (where possible).

File: src/assets/Generator.Shared/FormUrlEncodedContent.cs
Comment (AraHaan) [reply src/assets/Generator.Shared/FormUrlEncodedContent.cs:37]: What about the other 3 places where it is used?

File: src/AutoRest.CSharp/AutoRest.CSharp.csproj
Comment (AraHaan) [inline src/AutoRest.CSharp/AutoRest.CSharp.csproj:31]:
```suggestion
    <PackageReference Include="System.Memory" Version="4.5.4" />
```

File: src/assets/Generator.Shared/FormUrlEncodedContent.cs
Comment (AraHaan) [inline src/assets/Generator.Shared/FormUrlEncodedContent.cs:37]:
```suggestion
#if NET5_0_OR_GREATER || NETCOREAPP
```

File: src/assets/Generator.Shared/FormUrlEncodedContent.cs
Comment (AraHaan) [inline src/assets/Generator.Shared/FormUrlEncodedContent.cs:47]:
```suggestion
#if NET5_0_OR_GREATER || NETCOREAPP
```

File: src/assets/Generator.Shared/StringRequestContent.cs
Comment (AraHaan) [inline src/assets/Generator.Shared/StringRequestContent.cs:23]:
```suggestion
#if NET5_0_OR_GREATER || NETCOREAPP
```

File: src/assets/Generator.Shared/StringRequestContent.cs
Comment (AraHaan) [inline src/assets/Generator.Shared/StringRequestContent.cs:32]:
```suggestion
#if NET5_0_OR_GREATER || NETCOREAPP
```

File: src/AutoRest.CSharp/AutoRest.CSharp.csproj
Comment (AraHaan) [reply src/AutoRest.CSharp/AutoRest.CSharp.csproj]: Changed it to 4.5.4 for now.

**Comment by AraHaan** [comment]
I have made the requested changes.

File: src/assets/Generator.Shared/StringRequestContent.cs
Comment (jsquire) [inline src/assets/Generator.Shared/StringRequestContent.cs:23]: Should we just go with `NET6_0_OR_GREATER` here, since .NET 5 and .NET Core 3.1 are out of support?

Comment (AraHaan) [reply src/assets/Generator.Shared/StringRequestContent.cs:23]: Can yes, is there plans to drop those now on the azure sdk?

File: src/assets/Generator.Shared/StringRequestContent.cs
Comment (AraHaan) [inline src/assets/Generator.Shared/StringRequestContent.cs:32]:
```suggestion
#if NET6_0_OR_GREATER
```

File: src/assets/Generator.Shared/FormUrlEncodedContent.cs
Comment (AraHaan) [inline src/assets/Generator.Shared/FormUrlEncodedContent.cs:37]:
```suggestion
#if NET6_0_OR_GREATER
```

File: src/assets/Generator.Shared/FormUrlEncodedContent.cs
Comment (AraHaan) [inline src/assets/Generator.Shared/FormUrlEncodedContent.cs:47]:
```suggestion
#if NET6_0_OR_GREATER
```

**Comment by m-nash** [comment]
@AraHaan branch is out of date with feature/v3 can you please sync to the latest so the CI can run and validate?

**Comment by AraHaan** [comment]
I wish I could see why that smoketest is failing.

**Comment by archerzz** [comment]
> I wish I could see why that smoketest is failing.

@AraHaan Normally it's due to timeout:
> ECONNRESET

**Review by m-nash — APPROVED** [review]
Changes are fine but branch still says out of date with feature/v3 which will block merge

**Pull request merged**
